cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +127,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
+	case len(args) == 0:
+		// This should be reset only (but possibly for a region).
+		return c.handleZeroArgs()
+	case len(args) == 1:
+		// We want to get settings for the provided key.
+		return c.handleOneArg(args[0])
+	default: // case args > 1
+		// Specifying any positional args after a key=value pair is
+		// invalid input. So if we have more than one the input is almost
+		// certainly invalid, but in different possible ways.
+		return c.handleExtraArgs(args)
+	}
+
+}
+
+// parseResetKeys splits the keys provided to --reset after trimming any
+// leading or trailing comma. It then verifies that we haven't incorrectly
+// received any key=value pairs and finally sets the value(s) on c.resetKeys.
+func (c *defaultsCommand) parseResetKeys() error {
+	if len(c.reset) > 0 {
+		var resetKeys []string
+		for _, value := range c.reset {
+			keys := strings.Split(strings.Trim(value, ","), ",")
+			resetKeys = append(resetKeys, keys...)
 		}
-		for _, k := range args {
+
+		for _, k := range resetKeys {
 			if k == config.AgentVersionKey {
 				return errors.Errorf("%q cannot be reset", config.AgentVersionKey)
 			}
+			if strings.Contains(k, "=") {
+				return errors.Errorf(
+					`--reset accepts a single key "a" or comma delimited keys "a,b,c", received: %q`, k)
+			}
 		}
-		c.keys = args
-
-		c.action = c.resetDefaults
-		return nil
+		c.resetKeys = resetKeys
 	}
+	return nil
+}
 
-	if len(args) > 0 && strings.Contains(args[0], "=") {
-		// We're setting defaults.
-		options, err := keyvalues.Parse(args, true)
+// parseArgsForRegion parses args to check if the first arg is a region and
+// returns the appropriate remaining args.
+func (c *defaultsCommand) parseArgsForRegion(args []string) ([]string, error) {
+	var err error
+	if len(args) > 0 {
+		// determine if the first arg is cloud/region or region and return
+		// appropriate positional args.
+		args, err = c.parseCloudRegion(args)
 		if err != nil {
-			return errors.Trace(err)
-		}
-		c.values = make(attributes)
-		for k, v := range options {
-			if k == config.AgentVersionKey {
-				return errors.Errorf(`%q must be set via "upgrade-juju"`, config.AgentVersionKey)
-			}
-			c.values[k] = v
+			return nil, errors.Trace(err)
 		}
+	}
+	return args, nil
+}
 
-		c.action = c.setDefaults
-		return nil
+// parseCloudRegion examines args to see if the first arg is a cloud/region or
+// region. If not it returns the full args slice. If it is then it sets cloud
+// and/or region on the object and sends the remaining args back to the caller.
+func (c *defaultsCommand) parseCloudRegion(args []string) ([]string, error) {
+	var cloud, region string
+	cr := args[0]
+	// Must have no more than one slash and it must not be at the beginning or end.
+	if strings.Count(cr, "/") == 1 && !strings.HasPrefix(cr, "/") && !strings.HasSuffix(cr, "/") {
+		elems := strings.Split(cr, "/")
+		cloud, region = elems[0], elems[1]
+	} else {
+		region = cr
+	}
 
+	valid, err := c.validCloudRegion(cloud, region)
```

Comment:
Created a bug: pad.lv/1627162 made a note and added a short circuit return if region contains "=".